### PR TITLE
[ADD] l10n_in: Add IRN Number Field to Journal Entry and Update UOM Code

### PR DESCRIPTION
**Pull request by zepa-odoo**
This pull request updates the Indian localization module with the following improvements:

1. **UOM Code Update:**
   - Updated the Unit of Measure (UOM) code for compliance with the latest standards.

2. **IRN Number Integration:**
   - Added the `l10n_in_irn_number` field to the `account.move` model to store Invoice Reference Numbers (IRNs).
   - Updated invoice views to display the IRN field for Indian invoices.
   - Modified `account_edi_format` to fetch and store IRNs from the GST portal.

These changes enhance compliance with Indian GST regulations and streamline the invoicing process.

**Impact:**
- Automated IRN fetching and storage.

Related-Enterprise: https://github.com/odoo/enterprise/pull/67936

Task ID: 3551097

### CODE REVIEW

File: addons/account/views/account_journal_dashboard_view.xml [hunk 1]
```diff
@@ -309,28 +309,28 @@
...
-                                        <button type="object" name="action_create_new" class="o_invoice_new btn-secondary" groups="account.group_account_invoice">New</button>
+                        <div class="col mb-3 mb-sm-0 o_kanban_primary_left">
+                            <t t-if="journal_type == 'sale'">
+                                    <button type="object" name="action_create_new" class="btn btn-primary o_invoice_new" groups="account.group_account_invoice">
+                                        <span>New</span>
+                                    </button>
+                                <widget name="account_onboarding" t-if="journal_is_empty and dashboard['onboarding']"/>
+                            </t>
+                            <t t-if="journal_type == 'purchase'">
+                                <t t-if="!journal_is_empty">
```
Comment (yosuanicolaus) [inline addons/account/views/account_journal_dashboard_view.xml:320]: Since the `t-else` on the next element is a `div`, let's also refactor the only div child of this `t` element to here.

So it'll be `<div t-if="!journal_is_empty" class="d-flex"> <widget ...`

File: addons/l10n_in/models/res_partner.py [hunk 2]
```diff
@@ -72,3 +72,28 @@ def check_vat_in(self, vat):
...
+
+    @api.model
+    def l10n_in_get_partner_vals_by_vat(self, vat):
+        partner_details = self.env['res.partner'].read_by_vat(vat)
+        partner_data = partner_details[0] if partner_details else {}
+        if partner_data:
+            partner_gid = partner_data.get('partner_gid')
+            if partner_gid:
+                partner_data = self.env['res.partner'].enrich_company(company_domain=None, partner_gid=partner_gid, vat=partner_data.get('vat'))
+                partner_data = self.env['res.partner']._iap_replace_logo(partner_data)
```
Comment (yosuanicolaus) [inline addons/l10n_in/models/res_partner.py:84]: same here, no need for the `env[...]`

File: addons/l10n_in/models/res_partner.py [hunk 3]
```diff
@@ -72,3 +72,28 @@ def check_vat_in(self, vat):
...
+                'company_type': 'company',
+                'partner_gid': partner_gid,
+                'vat': partner_data.get('vat'),
+                'l10n_in_gst_treatment': 'regular',
+                'image_1920': partner_data.get('image_1920'),
+                'street': partner_data.get('street'),
+                'street2': partner_data.get('street2'),
+                'city': partner_data.get('city'),
+                'state_id': partner_data.get('state_id') and partner_data.get('state_id').get('id') or False,
+                'country_id': partner_data.get('country_id') and partner_data.get('country_id').get('id') or False,
```
Comment (yosuanicolaus) [inline addons/l10n_in/models/res_partner.py:96]: Were there any issue here previously? Why did you add all the ` and ...get.. or False` ?

If the `state_id` or `country_id` was not already an `id`, then you can do this

`partner_data.get('state_id', {}).get('id', False)`

But I think it should already be an `id`...

**Review by yosuanicolaus — CHANGES_REQUESTED** [review]
Hello :wave: is this PR ready for review? anyway, here's some comments

**Comment by zepa-odoo** [comment]
> Hello 👋 is this PR ready for review? anyway, here's some comments

Yes, it’s ready for review! 
I’ve addressed your comments as well. 
Thank you for your feedback!

File: addons/l10n_in/models/res_partner.py [hunk 4]
```diff
@@ -72,3 +72,28 @@ def check_vat_in(self, vat):
...
+                'company_type': 'company',
+                'partner_gid': partner_gid,
+                'vat': partner_data.get('vat'),
+                'l10n_in_gst_treatment': 'regular',
+                'image_1920': partner_data.get('image_1920'),
+                'street': partner_data.get('street'),
+                'street2': partner_data.get('street2'),
+                'city': partner_data.get('city'),
+                'state_id': partner_data.get('state_id', False),
+                'country_id': partner_data.get('country_id', False),
```
Comment (yosuanicolaus) [inline addons/l10n_in/models/res_partner.py:96]: Do we need the `False` default value here? Isn't `None` from `get` enough if we're just using this dict in `.create(...)` ?

**Review by yosuanicolaus — CHANGES_REQUESTED** [review]
Almost good. Don't forget to squash your commit at the end

File: addons/l10n_in/models/res_partner.py (hunk 4 above)
Comment (zepa-odoo) [reply addons/l10n_in/models/res_partner.py:96]: I will keep my the code back:

```python
'state_id': partner_data.get('state_id') and partner_data.get('state_id').get('id') or False,
```

The `read_by_vat` method returns `country_id` and `state_id` as dictionaries that include both `id` and `display_name`. Therefore, it's necessary to set the value to `False` if no data is retrieved.

For reference, you can check the code here where the values are set: Code Reference https://github.com/odoo/odoo/blob/5ad3315b260206be564c6ea08328d237bc656f5f/addons/partner_autocomplete/models/res_partner.py#L56

Comment (yosuanicolaus) [reply addons/l10n_in/models/res_partner.py]: I see, then let's improve it with my first suggestion then.
`'state_id': partner_data.get('state_id', {}).get('id', False)`

proof in python:
```py
>>> a = {'c': {'id': 123}}
>>> b = {}
>>> a.get('c', {}).get('id', False)
123
>>> b.get('c', {}).get('id', False)
False

Comment (zepa-odoo) [reply addons/l10n_in/models/res_partner.py]: Hello @yosuanicolaus ,
I have made the changes.

**Comment by jco-odoo** [comment]
@robodoo r+
